core/util_test.go

```diff
@@ -177,227 +310,296 @@ func (d *DrandTest2) GetBeacon(id string, round int, newGroup bool) (*chain.Beac
-func (d *DrandTest2) StartDrand(id string, catchup, newGroup bool) {
-	node := d.GetDrand(id, newGroup)
+func (d *DrandTestScenario) StartDrand(nodeAddress string, catchup, newGroup bool) {
+	node := d.GetMockNode(nodeAddress, newGroup)
 	dr := node.drand
+
 	// we load it from scratch as if the operator restarted its node
 	newDrand, err := LoadDrand(dr.store, dr.opts)
 	require.NoError(d.t, err)
 	node.drand = newDrand
-	fmt.Println("--- JUST BEFORE STARTBEACON---")
+
+	d.t.Logf("[drand] Start")
 	newDrand.StartBeacon(catchup)
-	fmt.Println("--- JUST AFTER STARTBEACON---")
+	d.t.Logf("[drand] Started")
 }
 
-func (d *DrandTest2) Now() time.Time {
+func (d *DrandTestScenario) Now() time.Time {
 	return d.clock.Now()
 }
 
-// MoveToTime sets the clock of all drands to the designated unix timestamp in
+// SetMockClock sets the clock of all drands to the designated unix timestamp in
 // seconds
-func (d *DrandTest2) MoveToTime(target int64) {
+func (d *DrandTestScenario) SetMockClock(t *testing.T, targetUnixTime int64) {
 	now := d.Now().Unix()
-	if now < target {
-		d.MoveTime(time.Duration(target-now) * time.Second)
+	if now < targetUnixTime {
+		d.AdvanceMockClock(t, time.Duration(targetUnixTime-now)*time.Second)
 	} else {
-		fmt.Println("ALREADY PASSED")
+		d.t.Logf("ALREADY PASSED")
 	}
+
+	t.Logf("Set genesis time: %d", d.Now().Unix())
 }
 
-// MoveTime advance the clock of all drand of the given duration
-func (d *DrandTest2) MoveTime(p time.Duration) {
+// AdvanceMockClock advances the clock of all drand by the given duration
+func (d *DrandTestScenario) AdvanceMockClock(t *testing.T, p time.Duration) {
 	for _, node := range d.nodes {
 		node.clock.Advance(p)
 	}
 	for _, node := range d.newNodes {
 		node.clock.Advance(p)
 	}
 	d.clock.Advance(p)
-	fmt.Printf(" --- MoveTime: new time is %d \n", d.Now().Unix())
-	time.Sleep(getSleepDuration())
 }
 
-// TestBeaconLength looks if the beacon chain on the given ids is of the
-// expected length
-func (d *DrandTest2) TestBeaconLength(length int, newGroup bool, ids ...string) {
+// CheckBeaconLength looks if the beacon chain on the given addresses is of the expected length
+func (d *DrandTestScenario) CheckBeaconLength(expectedLength int, newGroup bool, ids ...string) {
 	nodes := d.nodes
 	if newGroup {
 		nodes = d.resharedNodes
 	}
+
 	for _, id := range ids {
 		for _, node := range nodes {
+			var found bool
+
 			if node.addr != id {
 				continue
 			}
-			var found bool
-			for i := 0; i < 10; i++ {
+
+			retries := 10
+			for i := 0; i < retries; i++ {
 				inst := node.drand
-				if length != inst.beacon.Store().Len() {
-					time.Sleep(getSleepDuration())
-					continue
+
+				beaconLength := inst.beacon.Store().Len()
+				d.t.Logf("[beacon] length=%d retry=%d", beaconLength, i)
+
+				if expectedLength == beaconLength {
+					found = true
+					break
 				}
-				found = true
-				break
+
+				// TODO: Why a sleep here?
+				time.Sleep(getSleepDuration())
```

Comment:
IIRC it's because again of go routine. This function is not called after a callback for a new round has been inserted, is it ? So we dont know if the beacon was already inserted, so we try a few times. It's definitely up for some relifting as well this logic.

Reply:
Thanks!